activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb

```diff
@@ -0,0 +1,125 @@
+module ActiveRecord
+  module ConnectionAdapters
+    module MySQL
+      module DatabaseStatements
+        # Returns an ActiveRecord::Result instance.
+        def select_all(arel, name = nil, binds = [], preparable: nil)
+          result = if ExplainRegistry.collect? && prepared_statements
+            unprepared_statement { super }
+          else
+            super
+          end
+          @connection.next_result while @connection.more_results?
+          result
+        end
+
+        # Returns a record hash with the column names as keys and column values
+        # as values.
+        def select_one(arel, name = nil, binds = [])
+          arel, binds = binds_from_relation(arel, binds)
+          @connection.query_options.merge!(as: :hash)
```

Comment:
But how returning a hash instead of array in the result would be affected by the `result.each` call?